routing/directions_engine.cpp

```diff
@@ -0,0 +1,82 @@
+#include "routing/directions_engine.hpp"
+
+#include "base/assert.hpp"
+
+namespace
+{
+double constexpr KMPH2MPS = 1000.0 / (60 * 60);
+}  // namespace
+
+namespace routing
+{
+void IDirectionsEngine::CalculateTimes(IRoadGraph const & graph, vector<Junction> const & path,
+                                       Route::TTimes & times) const
+{
+  double const speedMPS = graph.GetMaxSpeedKMPH() * KMPH2MPS;
+
+  times.reserve(path.size());
+
+  double trackTimeSec = 0.0;
+  times.emplace_back(0, trackTimeSec);
+
+  m2::PointD prev = path[0].GetPoint();
+  for (size_t i = 1; i < path.size(); ++i)
+  {
+    m2::PointD const & curr = path[i].GetPoint();
+
+    double const lengthM = MercatorBounds::DistanceOnEarth(prev, curr);
+    trackTimeSec += lengthM / speedMPS;
+
+    times.emplace_back(i, trackTimeSec);
+
+    prev = curr;
+  }
+}
+
+bool IDirectionsEngine::ReconstructPath(IRoadGraph const & graph, vector<Junction> const & path,
+                                        vector<Edge> & routeEdges,
+                                        my::Cancellable const & cancellable) const
+{
+  routeEdges.clear();
+  if (path.size() <= 1)
+  {
+    ASSERT(false, (path.size()));
```

Comment:
According to code below we need to have a path of two points in ReconstructionPath so "if" is reasonable. As for ASSERT I'll try to remove the asserts and write tests.